elastic/apm-server
# Add "waitReady" preconditions for Elasticsearch clients

## Motivation/summary

When configured for standalone+data streams, we wait for the integration to be installed. This is now checked asynchronously, blocking Elasticsearch client requests rather than the server starting up. This allows events to be received, processed, and enqueued, but not indexed until the integration is installed. The tail-based sampling license check has been moved to this new wait loop, and uses the go-elasticsearch client to fetch the license instead of libbeat.

We add a "NewElasticsearchClient" function to beater.ServerParams, which should be used to construct `elasticsearch.Clients` wherever the client may be used to perform indexing operations. The elasticsearch package API has been simplified, and some unit tests removed as a result.

Package beater tests now have the option of publishing events using the standard libbeat elasticsearch output. We use this in the wait-for-integration test to check that events are not sent to Elasticsearch until the preconditions are met. We now deregister the pipeline registration callback when the server stops to fix tests which use the elasticsearch output, as otherwise the callback added in one test bleeds into those proceeding.

The "/" HTTP route handler now reports whether the server is "publish ready", i.e. it is ready to publish events.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

1. Run apm-server with `-E apm-server.data_streams.enabled=true`
2. Send some events, verify that they are not indexed
3. Install the integration package
4. Send some more events, verify that they are indexed

## Related issues

Closes https://github.com/elastic/apm-server/issues/5949
